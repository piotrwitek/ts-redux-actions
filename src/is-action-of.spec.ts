import { createAction, isActionOf } from '.';

// fixtures
const increment = createAction('INCREMENT');
const decrement = createAction('DECREMENT');
const add = createAction('ADD',
  (amount: number) => ({ type: 'ADD', payload: amount }),
);
const multiply = createAction('MULTIPLY',
  (amount: number) => ({ type: 'MULTIPLY', payload: amount }),
);
const divide = createAction('DIVIDE',
  (amount: number) => ({ type: 'DIVIDE', payload: amount }),
);

describe('isActionOf', () => {

  // TODO: #3
  // should error when missing argument
  // should error when passed invalid arguments
  // check object, empty array, primitives

  it('should succeed on action with correct type', () => {

    const isActionOfIncrement = isActionOf(increment);
    expect(isActionOfIncrement(increment())).toBeTruthy();

    const isActionOfIncrementOrAdd = isActionOf([increment, add]);
    expect(isActionOfIncrementOrAdd(increment())).toBeTruthy();
    expect(isActionOfIncrementOrAdd(add(2))).toBeTruthy();
  });

  it('should fail on action with incorrect type', () => {
    const isActionOfIncrement = isActionOf(increment);
    expect(isActionOfIncrement(add(2))).toBeFalsy();
    const isActionOfIncrement2 = isActionOf([increment]);
    expect(isActionOfIncrement2(add(2))).toBeFalsy();
  });

  it('should correctly assert type for EmptyAction', (done) => {
    const isActionOfIncrement = isActionOf(increment);
    const isActionOfIncrement2 = isActionOf([increment]);

    const action = { type: 'FALSE' };
    if (isActionOfIncrement(action)) {
      const a: { type: 'INCREMENT' } = action;
    } else if (isActionOfIncrement2(action)) {
      const a: { type: 'INCREMENT' } = action;
    } else {
      done();
    }
  });

  it('should correctly assert type for FluxStandardAction', (done) => {
    const isActionOfAdd = isActionOf(add);
    const isActionOfAdd2 = isActionOf([add]);

    const action = { type: 'FALSE' };
    if (isActionOfAdd(action)) {
      const a: { type: 'ADD', payload: number } = action;
    } else if (isActionOfAdd2(action)) {
      const a: { type: 'ADD', payload: number } = action;
    } else {
      done();
    }
  });

  it('should correctly assert array with one action', (done) => {
    const isActionOfOne = isActionOf([increment]);

    const action = { type: 'FALSE' };
    if (isActionOfOne(action)) {
      switch (action.type) {
        case 'INCREMENT': {
          const a: { type: 'INCREMENT' } = action;
          break;
        }
      }
    } else {
      done();
    }
  });

  it('should correctly assert array with two actions', (done) => {
    const isActionOfTwo = isActionOf([increment, decrement]);

    const action = { type: 'FALSE' };
    if (isActionOfTwo(action)) {
      switch (action.type) {
        case 'INCREMENT': {
          const a: { type: 'INCREMENT' } = action;
          break;
        }
        case 'DECREMENT': {
          const a: { type: 'DECREMENT' } = action;
          break;
        }
      }
    } else {
      done();
    }
  });

  it('should correctly assert array with three actions', (done) => {
    const isActionOfThree = isActionOf([increment, decrement, add]);

    const action = { type: 'FALSE' };
    if (isActionOfThree(action)) {
      switch (action.type) {
        case 'INCREMENT': {
          const a: { type: 'INCREMENT' } = action;
          break;
        }
        case 'DECREMENT': {
          const a: { type: 'DECREMENT' } = action;
          break;
        }
        case 'ADD': {
          const a: { type: 'ADD', payload: number } = action;
          break;
        }
      }
    } else {
      done();
    }
  });

  it('should correctly assert array with four actions', (done) => {
    const isActionOfFour = isActionOf([increment, decrement, add, multiply]);

    const action = { type: 'FALSE' };
    if (isActionOfFour(action)) {
      switch (action.type) {
        case 'INCREMENT': {
          const a: { type: 'INCREMENT' } = action;
          break;
        }
        case 'DECREMENT': {
          const a: { type: 'DECREMENT' } = action;
          break;
        }
        case 'ADD': {
          const a: { type: 'ADD', payload: number } = action;
          break;
        }
        case 'MULTIPLY': {
          const a: { type: 'MULTIPLY', payload: number } = action;
          break;
        }
      }
    } else {
      done();
    }
  });

  it('should correctly assert array with five actions', (done) => {
    const isActionOfFive = isActionOf([increment, decrement, add, multiply, divide]);

    const action = { type: 'FALSE' };
    if (isActionOfFive(action)) {
      switch (action.type) {
        case 'INCREMENT': {
          const a: { type: 'INCREMENT' } = action;
          break;
        }
        case 'DECREMENT': {
          const a: { type: 'DECREMENT' } = action;
          break;
        }
        case 'ADD': {
          const a: { type: 'ADD', payload: number } = action;
          break;
        }
        case 'MULTIPLY': {
          const a: { type: 'MULTIPLY', payload: number } = action;
          break;
        }
        case 'DIVIDE': {
          const a: { type: 'DIVIDE', payload: number } = action;
          break;
        }
      }
    } else {
      done();
    }
  });

});
