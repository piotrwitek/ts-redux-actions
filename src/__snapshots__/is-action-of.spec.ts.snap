// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`isActionOf should correctly assert for an array with 1 arg actual (type) should match snapshot 1`] = `"T.EmptyAction<\\"WITH_TYPE_ONLY\\">[]"`;

exports[`isActionOf should correctly assert for an array with 2 args actual (type) should match snapshot 1`] = `"(T.EmptyAction<\\"WITH_TYPE_ONLY\\"> | T.PayloadAction<\\"WITH_PAYLOAD\\", number>)[]"`;

exports[`isActionOf should correctly assert for an array with 3 args actual (type) should match snapshot 1`] = `"(T.EmptyAction<\\"WITH_TYPE_ONLY\\"> | T.PayloadAction<\\"WITH_PAYLOAD\\", number> | T.PayloadMetaAction<\\"WITH_PAYLOAD_META\\", number, string>)[]"`;

exports[`isActionOf should correctly assert for an array with 4 args actual (type) should match snapshot 1`] = `"(T.EmptyAction<\\"WITH_TYPE_ONLY\\"> | T.PayloadAction<\\"WITH_PAYLOAD\\", number> | ({ type: \\"WITH_MAPPED_PAYLOAD\\"; } & { payload: number; }) | T.PayloadMetaAction<\\"WITH_PAYLOAD_META\\", number, string>)[]"`;

exports[`isActionOf should correctly assert for an array with 5 args actual (type) should match snapshot 1`] = `"(T.EmptyAction<\\"WITH_TYPE_ONLY\\"> | T.PayloadAction<\\"WITH_PAYLOAD\\", number> | ({ type: \\"WITH_MAPPED_PAYLOAD\\"; } & { payload: number; }) | T.PayloadMetaAction<\\"WITH_PAYLOAD_META\\", number, string> | ({ type: \\"WITH_MAPPED_PAYLOAD_META\\"; } & { payload: number; meta: string; }))[]"`;

exports[`isActionOf should correctly assert type with "any" action action (type) should match snapshot 1`] = `"({ type: \\"WITH_MAPPED_PAYLOAD\\"; } & { payload: number; }) | ({ type: \\"WITH_MAPPED_PAYLOAD_META\\"; } & { payload: number; meta: string; })"`;

exports[`isActionOf should correctly assert type with "any" action action (type) should match snapshot 2`] = `"({ type: \\"WITH_MAPPED_PAYLOAD\\"; } & { payload: number; }) | ({ type: \\"WITH_MAPPED_PAYLOAD_META\\"; } & { payload: number; meta: string; })"`;

exports[`isActionOf should work with multiple action-creator args isActionOf([withTypeOnly, withPayload])(mappedPayloadAction) (type) should match snapshot 1`] = `"boolean"`;

exports[`isActionOf should work with multiple action-creator args isActionOf([withTypeOnly, withPayload])(payloadAction) (type) should match snapshot 1`] = `"boolean"`;

exports[`isActionOf should work with multiple action-creator args isActionOf([withTypeOnly, withPayload])(typeOnlyAction) (type) should match snapshot 1`] = `"boolean"`;

exports[`isActionOf should work with multiple action-creator args isActionOf([withTypeOnly, withPayload], mappedPayloadAction) (type) should match snapshot 1`] = `"boolean"`;

exports[`isActionOf should work with multiple action-creator args isActionOf([withTypeOnly, withPayload], payloadAction) (type) should match snapshot 1`] = `"boolean"`;

exports[`isActionOf should work with multiple action-creator args isActionOf([withTypeOnly, withPayload], typeOnlyAction) (type) should match snapshot 1`] = `"boolean"`;

exports[`isActionOf should work with single action-creator arg isActionOf([withTypeOnly])(payloadAction) (type) should match snapshot 1`] = `"boolean"`;

exports[`isActionOf should work with single action-creator arg isActionOf([withTypeOnly])(typeOnlyAction) (type) should match snapshot 1`] = `"boolean"`;

exports[`isActionOf should work with single action-creator arg isActionOf([withTypeOnly], payloadAction) (type) should match snapshot 1`] = `"boolean"`;

exports[`isActionOf should work with single action-creator arg isActionOf([withTypeOnly], typeOnlyAction) (type) should match snapshot 1`] = `"boolean"`;

exports[`isActionOf should work with single action-creator arg isActionOf(withTypeOnly)(payloadAction) (type) should match snapshot 1`] = `"boolean"`;

exports[`isActionOf should work with single action-creator arg isActionOf(withTypeOnly)(typeOnlyAction) (type) should match snapshot 1`] = `"boolean"`;

exports[`isActionOf should work with single action-creator arg isActionOf(withTypeOnly, payloadAction) (type) should match snapshot 1`] = `"boolean"`;

exports[`isActionOf should work with single action-creator arg isActionOf(withTypeOnly, typeOnlyAction) (type) should match snapshot 1`] = `"boolean"`;
